[NTOS:EX] Various fixes for Daylight Saving Time transitions

Doug-Lyons opened the pull request:
## Purpose

_Fix ExpSetTimeZoneInformation so that is is Daylight Savings Time aware._

JIRA issue: [CORE-19290](https://jira.reactos.org/browse/CORE-19290)

## Proposed changes

_Convert Eric Kohl's TimeZoneID code from ExRefreshTimeZoneInformation to a function._
_Use this new function in the ExpSetTimeZoneInformation function._

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
 
 /* FUNCTIONS ****************************************************************/
 
+static ULONG TimeZoneID(IN PLARGE_INTEGER TimeNow)
```
```suggestion
static ULONG TimeZoneID(_In_ PLARGE_INTEGER TimeNow)
```

Doug-Lyons replied:
Changed. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
+    if (ExpTimeZoneInfo.StandardDate.Month != 0 &&
+        ExpTimeZoneInfo.DaylightDate.Month != 0)
+    {
+        /* Get this years standard start time */
```
'years'?

Doug-Lyons replied:
Added semi-colon. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -347,13 +368,14 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
                                            TICKSPERMINUTE;
-    ExpTimeZoneId = TIME_ZONE_ID_STANDARD;
+
+    /* If Daylight Savings Time this subtracts an hour from the Time Zone Bias */
```
'subtracts an hour'?

Doug-Lyons replied:
Fixed. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
 LARGE_INTEGER ExpTimeZoneBias;
 ULONG ExpAltTimeZoneBias;
-ULONG ExpTimeZoneId;
+ULONG ExpTimeZoneId; // 0 = unknown, 1 = standard time, 2 = daylight time
```
Use defines or an enum?

HBelusca replied:
@Doug-Lyons Original code was using stuff like `TIME_ZONE_ID_STANDARD` etc. Why did you change this?

Doug-Lyons replied:
For only three (3) values, I considered this as not necessary. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
+                                        TimeNow,
+                                        TRUE))
+        {
+            DPRINT1("RtlCutoverTimeToSystemTime() for StandardDate failed!\n");
```
```suggestion
            DPRINT1("RtlCutoverTimeToSystemTime() for StandardDate failed\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
+                                        TimeNow,
+                                        TRUE))
+        {
+            DPRINT1("RtlCutoverTimeToSystemTime() for DaylightDate failed!\n");
```
```suggestion
            DPRINT1("RtlCutoverTimeToSystemTime() for DaylightDate failed\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
+            if ((TimeNow->QuadPart >= DaylightTime.QuadPart) &&
+                (TimeNow->QuadPart < StandardTime.QuadPart))
+            {
+                DPRINT("Daylight time!\n");
```
```suggestion
                DPRINT("Daylight time\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
+            }
+            else
+            {
+                DPRINT("Standard time!\n");
```
```suggestion
                DPRINT("Standard time\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
+            if ((TimeNow->QuadPart >= StandardTime.QuadPart) &&
+                (TimeNow->QuadPart < DaylightTime.QuadPart))
+            {
+                DPRINT("Standard time!\n");
```
```suggestion
                DPRINT("Standard time\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -22,14 +22,95 @@ RTL_TIME_ZONE_INFORMATION ExpTimeZoneInfo;
+            }
+            else
+            {
+                DPRINT("Daylight time!\n");
```
```suggestion
                DPRINT("Daylight time\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +306,13 @@ BOOLEAN
-        }
+    /* Set the Global Data for ExpTimeZoneBias and the Time Zone ID */
+    ExpTimeZoneId = TimeZoneID(CurrentBootTime);
+    DPRINT("ExpTimeZoneId is %d.\n", ExpTimeZoneId);
```
```suggestion
    DPRINT("ExpTimeZoneId is %d\n", ExpTimeZoneId);
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -336,6 +345,18 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
+
+    /* Set the Global Data for ExpTimeZoneBias and the Time Zone ID */
+    ExpTimeZoneId = TimeZoneID(&SystemTime);
+    DPRINT("ExpTimeZoneId is %d.\n", ExpTimeZoneId);
```
```suggestion
    DPRINT("ExpTimeZoneId is %d\n", ExpTimeZoneId);
```

Doug-Lyons replied:
Changed. Thanks.

Doug-Lyons commented:
@HBelusca, regarding your question about 'Why did you change this?', I changed none of Eric Kohl's original code in this respect. I had used a hard-coded '2' in ExpSetTimeZoneInformation for a compare, but I have now changed this to use 'TIME_ZONE_ID_DAYLIGHT'. Also, I moved the comments regarding the numbers to a longer comment in the function's header. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -336,6 +356,18 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
+    while (SystemTime.HighPart != SharedUserData->SystemTime.High2Time);
+
+    /* Set the Global Data for ExpTimeZoneBias and the Time Zone ID */
+    ExpTimeZoneId = TimeZoneID(&SystemTime);
```
The results of this function are used in ZwQuerySystemInformation(SystemCurrentTimeZoneInformation,...) which needs to return an NTSTATUS. So on fail, I now return STATUS_UNSUCCESSFUL. Thanks again.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +317,13 @@ BOOLEAN
-        ExpTimeZoneId = TIME_ZONE_ID_UNKNOWN;
-    }
+    /* Set the Global Data for ExpTimeZoneBias and the Time Zone ID */
+    ExpTimeZoneId = TimeZoneID(CurrentBootTime);
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+    {
+        ExpTimeZoneId = TIME_ZONE_ID_UNKNOWN;
+    }
+    return ExpTimeZoneId;
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+    }
+    else
+    {
+        ExpTimeZoneId = TIME_ZONE_ID_UNKNOWN;
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+            else
+            {
+                DPRINT("Daylight time\n");
+                ExpTimeZoneId = TIME_ZONE_ID_DAYLIGHT;
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+                (TimeNow->QuadPart < DaylightTime.QuadPart))
+            {
+                DPRINT("Standard time\n");
+                ExpTimeZoneId = TIME_ZONE_ID_STANDARD;
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+            else
+            {
+                DPRINT("Standard time\n");
+                ExpTimeZoneId = TIME_ZONE_ID_STANDARD;
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+                (TimeNow->QuadPart < StandardTime.QuadPart))
+            {
+                DPRINT("Daylight time\n");
+                ExpTimeZoneId = TIME_ZONE_ID_DAYLIGHT;
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
 
+/*++
+ * This function sets the following Global Variables:
+ * ExpTimeZoneID
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+ *     1 = TIME_ZONE_ID_STANDARD
+ *     2 = TIME_ZONE_ID_DAYLIGHT
+ *--*/
+static ULONG TimeZoneID(_In_ PLARGE_INTEGER TimeNow)
```
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,98 @@ ULONG ExpTimerResolutionCount = 0;
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        return FALSE;
```
Changed now. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -347,13 +387,15 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
+
+    /* If Daylight Savings Time then add the DayLightBias to the Time Zone Bias */
+    if (ExpTimeZoneId == TIME_ZONE_ID_DAYLIGHT)
+
```
```suggestion
```

Doug-Lyons replied:
Changed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -334,8 +357,25 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
 {
     LARGE_INTEGER LocalTime, SystemTime, OldTime;
     TIME_FIELDS TimeFields;
+    NTSTATUS Status = STATUS_UNSUCCESSFUL; // Allow easy failure return change
```
```suggestion
```

Doug-Lyons replied:
Changed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -334,8 +357,25 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
+    /* Set the Global Data for ExpTimeZoneBias and ExpTimeZoneId */
+    if (!ExpGetTimeZoneId(&SystemTime, &ExpTimeZoneId))
+    {
+        DPRINT1("ExpSetTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        return Status;
```
```suggestion
        DPRINT1("ExpGetTimeZoneId() failed\n");
        return STATUS_UNSUCCESSFUL;
```
Not sure the dprint1 is that useful though.... (it already emits dprints when the functions it calls fail)

Doug-Lyons replied:
Changed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -347,13 +387,15 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
     HalQueryRealTimeClock(&TimeFields);
     RtlTimeFieldsToTime(&TimeFields, &LocalTime);
 
-    /* FIXME: Calculate transition dates */
```
And what's that FIXME about?

Doug-Lyons replied:
This has been in the code since Eric Kohl originally wrote this section of the code on 12/1/2004 in R11885 and commit 019c9076 [019c9076eeb2d3696c20160424d4a86a5430efc6]. I believe that he placed it here as a placeholder to note that he was not calculating the Daylight Saving Time Start and Standard Time Start times and accounting for them. I removed the comment, because with this new revision these are not taken into account. Thanks.

SergeGautherie replied:
> on 12/1/2004 in R11885 and commit [019c907](https://github.com/reactos/reactos/commit/019c9076eeb2d3696c20160424d4a86a5430efc6)

Rather df74513933549485017561ea18e551c8ddcf8c93 (r11867).

Doug-Lyons replied:
Sorry, I looked at the left side of the diff instead of the right. @SergeGautherie, you are correct. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,100 @@ ULONG ExpTimerResolutionCount = 0;
+ * ExpTimeZoneInfo
+ * ExpTimeZoneBias
+ *
+ * It returns ExpTimeZoneBias as follows:
+ *     0 = TIME_ZONE_ID_UNKNOWN
+ *     1 = TIME_ZONE_ID_STANDARD
+ *     2 = TIME_ZONE_ID_DAYLIGHT
```
This comment does not make sense.

Doug-Lyons replied:
Hopefully improved now. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,100 @@ ULONG ExpTimerResolutionCount = 0;
+ *     1 = TIME_ZONE_ID_STANDARD
+ *     2 = TIME_ZONE_ID_DAYLIGHT
+ *--*/
+static BOOLEAN
```
```suggestion
static
BOOLEAN
```

Doug-Lyons replied:
Changed. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +319,14 @@ BOOLEAN
+    if (!ExpGetTimeZoneId(CurrentBootTime, &ExpTimeZoneId))
         return FALSE;
-    }
+    DPRINT("ExpTimeZoneId is %d\n", ExpTimeZoneId);
```
```suggestion
    DPRINT("ExpTimeZoneId is %lu\n", ExpTimeZoneId);
```

Doug-Lyons replied:
Changed. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -334,8 +357,25 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
+        DPRINT1("ExpSetTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        return Status;
+    }
+    DPRINT("ExpTimeZoneId is %d\n", ExpTimeZoneId);
```
```suggestion
    DPRINT("ExpTimeZoneId is %lu\n", ExpTimeZoneId);
```

Doug-Lyons replied:
Changed. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -468,6 +527,15 @@ NtSetSystemTime(IN PLARGE_INTEGER SystemTime,
+    /* Read time zone information from the registry */
+    Status = RtlQueryTimeZoneInformation(&TimeZoneInformation);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+    }
+
+    ExpSetTimeZoneInformation(&TimeZoneInformation);
```
```suggestion
        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
    else
        Status = ExpSetTimeZoneInformation(&TimeZoneInformation);
```

Doug-Lyons replied:
I think that we want to do this even if we receive a failure code from RtlQueryTimeZoneInformation. It's just the best that we can do here. Thanks.

SergeGautherie replied:
Using an uninitialized `TimeZoneInformation` doesn't look good to me.

Doug-Lyons replied:
@SergeGautherie, We set the TimeZoneInformation values to all zeroes here:

RTL_TIME_ZONE_INFORMATION TimeZoneInformation = {0};

Then we make the call to RtlQueryTimieZoneInformation like this:

Status = RtlQueryTimeZoneInformation(&TimeZoneInformation);

RtlQueryTimeZoneInformation tries to fill this struct with values from the registry.
If it fails, then the original zeroes just remain in place.
Proceeding with zeroes in the TimeZoneInformation structure basically just leaves things alone.
So basically the DaylightDate, DaylightBias, StandardDate, and StandardBias members are all zero.
So changes occur in the code below and no harm is done.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,115 @@ ULONG ExpTimerResolutionCount = 0;
+ * ExpTimeZoneBias
+ * ExpTimeZoneId
+ *
+ * It returns *TimeZoneId as follows:
```
```suggestion
 * It sets *TimeZoneId as follows:
```

Doug-Lyons replied:
Improved based on suggestion. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,115 @@ ULONG ExpTimerResolutionCount = 0;
+ * ExpTimeZoneId
+ *
+ * It returns *TimeZoneId as follows:
+ *     0 = TIME_ZONE_ID_UNKNOWN
+ *     1 = TIME_ZONE_ID_STANDARD
+ *     2 = TIME_ZONE_ID_DAYLIGHT
```
Why should the actual numerical values matter?

Doug-Lyons replied:
If someone wanted to print out a single value, then could decode it based on this table.

SergeGautherie replied:
They should look at the defines directly:
https://github.com/search?q=repo%3Areactos%2Freactos%20TIME_ZONE_ID_STANDARD&type=code
Duplicating the values (again) is not good.

Doug-Lyons replied:
@SergeGautherie, I have spent many hours lately modifying and trouble-shooting this code.
In doing so, I have liberally added statements such as the following to my testing code:
```
    DPRINT1("*TimeZoneId is %d\n", *TimeZoneId);
    DPRINT1("ExpTimeZoneId is %d\n", ExpTimeZoneId);
```
And, I know that technically the print specifier above is incorrect.
So knowing and using these values by their absolute numeric values was very helpful to me.
Therefore, I would hope to help any others that come to this code in the future.
But in the interest of keeping peace here, I will delete this section of code.

Doug-Lyons replied:
Code section deleted. Thanks.

SergeGautherie commented on ntoskrnl/ex/time.c:
```diff
@@ -468,6 +528,32 @@ NtSetSystemTime(IN PLARGE_INTEGER SystemTime,
+    ExpTimeZoneIdSave = ExpTimeZoneId;
+    ExpSetTimeZoneInformation(&TimeZoneInformation);
+
+    if (ExpTimeZoneIdSave != ExpTimeZoneId)
```
```suggestion
    if (ExpTimeZoneId != ExpTimeZoneIdSave)
```

Doug-Lyons replied:
Changed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,115 @@ ULONG ExpTimerResolutionCount = 0;
+
+    if (!InitialRun)
+    {
+        /* Adjust TimeNow for Time Zone Bias to use UTC for compares */
```
"compares" --> "comparisons" ?

Doug-Lyons replied:
Changed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -468,6 +528,32 @@ NtSetSystemTime(IN PLARGE_INTEGER SystemTime,
+    Status = RtlQueryTimeZoneInformation(&TimeZoneInformation);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlQueryTimeZoneInformation failed (Status 0x%08lx)\n", Status);
```
As Serge pointed, here TimeZoneInformation is very certainly still non-initialized, so using it below is dangerous.
What do you want to do if querying the timezone info fails ?

Doug-Lyons replied:
I intend to initialize this in its definition like the following:
```
    RTL_TIME_ZONE_INFORMATION TimeZoneInformation = { 0 };
```
Testing for zeroes (0) is done in some of this code to determine the code patch already.
Updates to follow shortly.

Doug-Lyons replied:
Now done. Thanks.

HBelusca replied:
@Doug-Lyons I've just noticed that in the current PR this suggestion wasn't still applied; see my comments below.

Doug-Lyons replied:
@HBelusca, Yes. I changed my mind and initialized it in the line numbers 471 and 472 above so that I could explicitly do it there and put a comment so that it would not be missed, but now I have restored it to what it was before because it is simpler this way. Thanks.

Doug-Lyons commented:
Modified code to explicitly zero all fields of TimeZoneInformation as requested by @HBelusca.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,113 @@ ULONG ExpTimerResolutionCount = 0;
+        return FALSE;
+    }
+
+    if (!InitialRun)
+    {
+        /* Adjust TimeNow for Time Zone Bias to use UTC for comparisons */
+        TimeNow->QuadPart -= (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;
+    }
+    else
+    {
+        InitialRun = FALSE;
+    }
+
+    /* Get the default bias */
+    ExpTimeZoneBias.QuadPart = (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;
```
It may be of interest to get this default bias first, then do the InitialRun check and adjust the TimeNow with that bias: something like this:
```suggestion
    /* Get the default bias */
    ExpTimeZoneBias.QuadPart = (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;

    if (!InitialRun)
    {
        /* Adjust TimeNow for Time Zone Bias to use UTC for comparisons */
        TimeNow->QuadPart -= ExpTimeZoneBias.QuadPart;
    }
    else
    {
        InitialRun = FALSE;
    }
```

HBelusca replied:
By the way, note that the LARGE_INTEGER variable passed via pointer : TimeNow, is also modified. I'm not sure this is correct.
You may have to locally capture its value in order to be able to internally use/modify it inside this function, without affecting what the caller of this function sees once the function returns.
By "locally capturing" the value, I mean, doing something like this:
```c
LARGE_INTEGER LocalTimeNow = *TimeNow;
...
LocalTimeNow.QuadPart -= ExpTimeZoneBias.QuadPart;
...
// Use LocalTimeNow instead of TimeNow in the rest of the code in this function.
```

Doug-Lyons replied:
Changes made as suggested. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -412,6 +470,9 @@ NtSetSystemTime(IN PLARGE_INTEGER SystemTime,
         return STATUS_NOT_IMPLEMENTED;
     }
 
+    // Initialize all fields of TimeZoneInformation to zeroes
+    memset(&TimeZoneInformation, 0, sizeof(TimeZoneInformation));
+
```
```suggestion
```

Doug-Lyons replied:
Removed.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -401,6 +457,8 @@ NtSetSystemTime(IN PLARGE_INTEGER SystemTime,
     TIME_FIELDS TimeFields;
     KPROCESSOR_MODE PreviousMode = ExGetPreviousMode();
     NTSTATUS Status = STATUS_SUCCESS;
+    RTL_TIME_ZONE_INFORMATION TimeZoneInformation;
```
```suggestion
    RTL_TIME_ZONE_INFORMATION TimeZoneInformation = {0};
```

Doug-Lyons replied:
Changed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,113 @@ ULONG ExpTimerResolutionCount = 0;
+ * ExpTimeZoneBias
+ * ExpTimeZoneId
+ *
+ * This variable is modified and returned as a result of this function
```
I think this is now evident, via the `_Out_` annotation.

Doug-Lyons replied:
Removed unneeded comments. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,110 @@ ULONG ExpTimerResolutionCount = 0;
+        *TimeZoneId = TIME_ZONE_ID_UNKNOWN;
+    }
+
+    DPRINT("ExpTimeZoneId is %lu\n", ExpTimeZoneId);
```
```suggestion
```

Doug-Lyons replied:
Removed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,110 @@ ULONG ExpTimerResolutionCount = 0;
+ * If successful, this function sets the following Global Variables:
+ * ExpTimeZoneInfo
+ * ExpTimeZoneBias
+ * ExpTimeZoneId
```
```suggestion
```

Doug-Lyons replied:
Removed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,110 @@ ULONG ExpTimerResolutionCount = 0;
+        /* Get this year's standard start time */
+        if (!RtlCutoverTimeToSystemTime(&ExpTimeZoneInfo.StandardDate,
+                                        &StandardTime,
+                                        TimeNow,
```
```suggestion
                                        &LocalTimeNow,
```

Doug-Lyons replied:
Changed. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,110 @@ ULONG ExpTimerResolutionCount = 0;
+        /* Get this year's daylight start time */
+        if (!RtlCutoverTimeToSystemTime(&ExpTimeZoneInfo.DaylightDate,
+                                        &DaylightTime,
+                                        TimeNow,
```
```suggestion
                                        &LocalTimeNow,
```

Doug-Lyons replied:
Changed. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,110 @@ ULONG ExpTimerResolutionCount = 0;
 /* FUNCTIONS ****************************************************************/
 
+/*++
+ * If successful, this function sets the following Global Variables:
```
Revised. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +329,17 @@ BOOLEAN
+    if (!ExpGetTimeZoneId(CurrentBootTime, &ExpTimeZoneId))
     {
-        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        DPRINT1("ExpTimeZoneId failed returning %lu\n", ExpTimeZoneId);
```
Changed to ```DPRINT1("ExpGetTimeZoneId() failed\n");```. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,110 @@ ULONG ExpTimerResolutionCount = 0;
+ * If successful, this function sets the following Global Variables:
+ * ExpTimeZoneInfo
+ * ExpTimeZoneBias
+ * ExpTimeZoneId
```
This line has now been removed and should no longer be referenced. Thanks.

Doug-Lyons commented on ntoskrnl/ex/time.c:
```diff
@@ -468,6 +523,32 @@ NtSetSystemTime(IN PLARGE_INTEGER SystemTime,
+    {
+        /* Going from DT to ST or vice versa we need to repeat this */
+        DPRINT("Daylight Time and Standard Time are switching\n");
+        Status = RtlQueryTimeZoneInformation(&TimeZoneInformation);
```
Removed call and associated code. Thanks.

JoachimHenze reviewed (approved):
Approved. But let's please have a testbot run before doing the merge.

tkreuzer commented on ntoskrnl/ex/time.c:
```diff
@@ -336,6 +370,22 @@ ExpSetTimeZoneInformation(PRTL_TIME_ZONE_INFORMATION TimeZoneInformation)
     TIME_FIELDS TimeFields;
     DPRINT("ExpSetTimeZoneInformation() called\n");
 
+    /* Get the Shared User Data System Time into the local SystemTime */
+    do
+    {
+        SystemTime.HighPart = SharedUserData->SystemTime.High1Time;
+        SystemTime.LowPart = SharedUserData->SystemTime.LowPart;
+    }
+    while (SystemTime.HighPart != SharedUserData->SystemTime.High2Time);
+
```
Use KeQuerySystemTime instead.

Doug-Lyons replied:
Revised to use KeQuerySystemTime. Thanks.

binarymaster commented:
Note that I improved PR title wording a bit, please take that into account on merging.

tkreuzer commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +327,17 @@ BOOLEAN
+    if (!ExpGetTimeZoneId(CurrentBootTime, &ExpTimeZoneId))
     {
-        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        DPRINT1("ExpTimeZoneId failed\n");
```
"ExpGetTimeZoneId failed"

Doug-Lyons replied:
Fixed. Thanks.

tkreuzer commented on ntoskrnl/ex/time.c:
```diff
@@ -401,6 +452,8 @@ NtSetSystemTime(IN PLARGE_INTEGER SystemTime,
     KPROCESSOR_MODE PreviousMode = ExGetPreviousMode();
     NTSTATUS Status = STATUS_SUCCESS;
+    RTL_TIME_ZONE_INFORMATION TimeZoneInformation = { 0 };
+    ULONG ExpTimeZoneIdSave;
```
Please remove the `Exp` prefix, that looks like it would be a global variable.

Doug-Lyons replied:
Removed the ```Exp``` variable prefix. Thanks.

tkreuzer commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,108 @@ ULONG ExpTimerResolutionCount = 0;
+ *--*/
+static
+BOOLEAN
+ExpGetTimeZoneId(
```
The function is called Exp**Get**TimeZoneId, which implies that it doesn't modify any global state, but it does. And it even does on failure. I suggest instead of modifying ExpTimeZoneBias in this function, to just return it in a 3rd parameter and let the caller set the global variable on success.

Doug-Lyons replied:
Added third parameter to return NewTimeZoneBias. Thanks.

tkreuzer commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,108 @@ ULONG ExpTimerResolutionCount = 0;
+
+    /* Get the default bias */
+    ExpTimeZoneBias.QuadPart = (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;
+    if (!InitialRun)
+    {
+        /* Adjust TimeNow for Time Zone Bias to use UTC for comparisons */
+        LocalTimeNow.QuadPart -= ExpTimeZoneBias.QuadPart;
+    }
+    else
+    {
+        InitialRun = FALSE;
+    }
```
I don't really like the use of a static local and different behavior based on when the function is called. Instead I suggest to modify the initial caller, so that it adds ExpTimeZoneBias to the time it passes to this function, so that the logic is right again.

Doug-Lyons replied:
Removed static ```InitialRun``` variable and this whole section of code. Replaced it with calculating the Time Zone Bias before calling this function and accounting for it before making the call, so it comes in with the input parameters.

Doug-Lyons commented:
@tkreuzer, Thanks for your review. I believe that I have taken care of fixing all of your comments and concerns. Can you please check and confirm? Thanks again.

Doug-Lyons commented:
@HBelusca, Can you please review these latest changes and see if you can provide your approval here? Thanks in advance.

Doug-Lyons commented:
Testbot Results:

refs/pull/5868/merge on top of 0.4.15-dev-8023-g30226f8

KVM:  https://reactos.org/testman/compare.php?ids=95261,95264 LGTM
VBox: https://reactos.org/testman/compare.php?ids=95260,95265 LGTM

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +319,23 @@ BOOLEAN
+    ExpTimeZoneBias = NewTimeZoneBias;
 
-    /* Change it for user-mode applications */
+    /* Change SharedUserData->TimeZoneBias for user-mode applications */
```
(Since I cannot comment at the correct place of code, I'm doing it there):
@Doug-Lyons , could you also replace the `/* Convert system time from local time to UTC */` do-while loop below, that retrieves `SharedUserData->SystemTime` into `CurrentTime` below (old lines 316-320 / new lines 348-352) with a `KeQuerySystemTime(&CurrentTime);` too?

In addition I think that the resetting of the `SharedUserData->SystemTime` after that loop should be changed so that the ordering of setting the High1/2Time and LowPart should respect the ordering as done also here:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/clock.c;hb=8e799e2d3eeaf350acda68700577671b405580d9#l54

Doug-Lyons replied:
@HBelusca, I have not been able to figure out how to change this section of code to use ```KeQuerySystemTime(&CurrentTime);```. To make this match the replaced code, it seems that for the parameter I need to use ```&CurrentTime.u```, but this produces a compile time error that the parameter needs to be a ```PLARGE_INTEGER```. This variable type is complex and uses ```union``` and ```struct``` together and both of these are foreign to me.
If you will give me complete line-by-line code changes, then I will try this again. Sorry, but this is above me for now. Thanks.

Doug-Lyons replied:
I have changed the ordering to High2Time, LowPart, and then High1Time in two places. Please confirm that this is what you wanted. Thanks.

HBelusca replied:
> @HBelusca, I have not been able to figure out how to change this section of code to use `KeQuerySystemTime(&CurrentTime);`. To make this match the replaced code, it seems that for the parameter I need to use `&CurrentTime.u`, [...]

No, since [`KeQuerySystemTime()` takes a pointer to `LARGE_INTEGER` for input/output](https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/clock.c;hb=8e799e2d3eeaf350acda68700577671b405580d9#l184), and since `CurrentTime` is already a `LARGE_INTEGER`, then the suggested `KeQuerySystemTime(&CurrentTime);` should compile OK and works as expected.

HBelusca replied:
> I have changed the ordering to High2Time, LowPart, and then High1Time in two places. Please confirm that this is what you wanted. Thanks.

I confirm your changes are OK.

Doug-Lyons replied:
OK. now with the ```KeQuerySystemTime(&CurrentTime);```. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +319,23 @@ BOOLEAN
-    LARGE_INTEGER CurrentTime;
-    NTSTATUS Status;
+    LARGE_INTEGER CurrentTime, NewTimeZoneBias;
+    BOOLEAN result;
```
`result` --> `Success`

Doug-Lyons replied:
Done. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -225,85 +319,23 @@ BOOLEAN
-    if (!NT_SUCCESS(Status))
+    DPRINT("ExRefreshTimeZoneInformation\n");
+
+    /* Set the Global Data for ExpTimeZoneBias and the Time Zone ID */
```
```suggestion
    /* Set the global data for ExpTimeZoneBias and the Time Zone ID */
```

Doug-Lyons replied:
Done. Thanks.

HBelusca commented on ntoskrnl/ex/time.c:
```diff
@@ -332,10 +364,44 @@ ExRefreshTimeZoneInformation(IN PLARGE_INTEGER CurrentBootTime)
-    LARGE_INTEGER LocalTime, SystemTime, OldTime;
+    LARGE_INTEGER LocalTime, SystemTime, OldTime, NewTimeZoneBias;
     TIME_FIELDS TimeFields;
+    BOOLEAN result;
```
`result` ---> `Success`

Doug-Lyons replied:
Done. Thanks.

HBelusca reviewed (approved):
minor fixes (re. the query system time) that need to be applied before merging the PR, but otherwise, OK.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -332,10 +360,44 @@ ExRefreshTimeZoneInformation(IN PLARGE_INTEGER CurrentBootTime)
+    Success = ExpGetTimeZoneId(&LocalTimeNow, &ExpTimeZoneId, &NewTimeZoneBias);
+    if (!Success)
+    {
+        DPRINT1("ExpGetTimeZoneId() failed\n");
```
```suggestion
        DPRINT1("ExpGetTimeZoneId failed\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -332,10 +360,44 @@ ExRefreshTimeZoneInformation(IN PLARGE_INTEGER CurrentBootTime)
+    Status = RtlQueryTimeZoneInformation(&ExpTimeZoneInfo);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
```
```suggestion
        DPRINT1("RtlQueryTimeZoneInformation failed (Status 0x%08lx)\n", Status);
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,100 @@ ULONG ExpTimerResolutionCount = 0;
+    Status = RtlQueryTimeZoneInformation(&ExpTimeZoneInfo);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
```
```suggestion
        DPRINT1("RtlQueryTimeZoneInformation failed (Status 0x%08lx)\n", Status);
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,100 @@ ULONG ExpTimerResolutionCount = 0;
+                                        &LocalTimeNow,
+                                        TRUE))
+        {
+            DPRINT1("RtlCutoverTimeToSystemTime() for StandardDate failed\n");
```
```suggestion
            DPRINT1("RtlCutoverTimeToSystemTime for StandardDate failed\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze commented on ntoskrnl/ex/time.c:
```diff
@@ -30,6 +30,100 @@ ULONG ExpTimerResolutionCount = 0;
+                                        &LocalTimeNow,
+                                        TRUE))
+        {
+            DPRINT1("RtlCutoverTimeToSystemTime() for DaylightDate failed\n");
```
```suggestion
            DPRINT1("RtlCutoverTimeToSystemTime for DaylightDate failed\n");
```

Doug-Lyons replied:
Changed. Thanks.

JoachimHenze reviewed (commented):
Suggestions for consistency and to save a few bytes.

Doug-Lyons commented:
Thanks for everyone's help with this and thanks for the approvals from @tkreuzer, @HBelusca, and @JoachimHenze.
